### PR TITLE
Add Roles contract

### PR DESCRIPTION
**Pull request by sunbreak1211**

### CODE REVIEW

File: test/AllocatorRoles.t.sol [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+		assertEq(roles.publicActions(address(authed), bytes4(keccak256("exec()"))), 1);
+		assertTrue(roles.canCall(address(this), address(authed), bytes4(keccak256("exec()"))));
+		authed.exec();
+
+		roles.setPublicAction(address(authed), bytes4(keccak256("exec()")), false);
+		assertEq(roles.publicActions(address(authed), bytes4(keccak256("exec()"))), 0);
+		assertTrue(!roles.canCall(address(this), address(authed), bytes4(keccak256("exec()"))));
+		vm.expectRevert("AuthedMock/not-authorized");
+		authed.exec();
+	}
```
Comment (telome) [inline test/AllocatorRoles.t.sol]: This file seems to have a mix of tabs and 4-spaces 

```suggestion
    AllocatorRoles roles;
    AuthedMock     authed;

    function setUp() public {
        roles = new AllocatorRoles();
        authed = new AuthedMock(address(roles));
    }

    function testAuth() public {
        checkAuth(address(roles), "AllocatorRoles");
    }

    function testModifiers() public {
        bytes4[] memory authedMethods = new bytes4[](3);
        authedMethods[0] = roles.setUserRole.selector;
        authedMethods[1] = roles.setPublicAction.selector;
        authedMethods[2] = roles.setRoleAction.selector;

        vm.startPrank(address(0xBEEF));
        checkModifier(address(roles), "AllocatorRoles/not-authorized", authedMethods);
        vm.stopPrank();
    }

    function testBasics() public {
        uint8 admin_role = 0;
        uint8 mod_role = 1;
        uint8 user_role = 2;

        assertTrue(!roles.hasUserRole(address(this), admin_role));
        assertTrue(!roles.hasUserRole(address(this), mod_role));
        assertTrue(!roles.hasUserRole(address(this), user_role));
        assertEq32(bytes32(hex"0000000000000000000000000000000000000000000000000000000000000000"), roles.userRoles(address(this)));

        roles.setUserRole(address(this), admin_role, true);

        assertTrue( roles.hasUserRole(address(this), admin_role));
        assertTrue(!roles.hasUserRole(address(this), mod_role));
        assertTrue(!roles.hasUserRole(address(this), user_role));
        assertEq32(bytes32(hex"0000000000000000000000000000000000000000000000000000000000000001"), roles.userRoles(address(this)));

        assertTrue(!roles.canCall(address(this), address(authed), bytes4(keccak256("exec()"))));
        vm.expectRevert("AuthedMock/not-authorized");
        authed.exec();

        roles.setRoleAction(admin_role, address(authed), bytes4(keccak256("exec()")), true);

        assertTrue(roles.canCall(address(this), address(authed), bytes4(keccak256("exec()"))));
        authed.exec();
        assertTrue(authed.flag());

        roles.setRoleAction(admin_role, address(authed), bytes4(keccak256("exec()")), false);
        assertTrue(!roles.canCall(address(this), address(authed), bytes4(keccak256("exec()"))));
        vm.expectRevert("AuthedMock/not-authorized");
        authed.exec();

        roles.setUserRole(address(this), mod_role, true);

        assertTrue( roles.hasUserRole(address(this), admin_role));
        assertTrue( roles.hasUserRole(address(this), mod_role));
        assertTrue(!roles.hasUserRole(address(this), user_role));
        assertEq32(bytes32(hex"0000000000000000000000000000000000000000000000000000000000000003"), roles.userRoles(address(this)));

        roles.setUserRole(address(this), user_role, true);

        assertTrue( roles.hasUserRole(address(this), admin_role));
        assertTrue( roles.hasUserRole(address(this), mod_role));
        assertTrue( roles.hasUserRole(address(this), user_role));
        assertEq32(bytes32(hex"0000000000000000000000000000000000000000000000000000000000000007"), roles.userRoles(address(this)));

        roles.setUserRole(address(this), mod_role, false);

        assertTrue( roles.hasUserRole(address(this), admin_role));
        assertTrue(!roles.hasUserRole(address(this), mod_role));
        assertTrue( roles.hasUserRole(address(this), user_role));
        assertEq32(bytes32(hex"0000000000000000000000000000000000000000000000000000000000000005"), roles.userRoles(address(this)));
    }

    function testPublicActions() public {
        assertEq(roles.publicActions(address(authed), bytes4(keccak256("exec()"))), 0);
        assertTrue(!roles.canCall(address(this), address(authed), bytes4(keccak256("exec()"))));
        vm.expectRevert("AuthedMock/not-authorized");
        authed.exec();

        roles.setPublicAction(address(authed), bytes4(keccak256("exec()")), true);
        assertEq(roles.publicActions(address(authed), bytes4(keccak256("exec()"))), 1);
        assertTrue(roles.canCall(address(this), address(authed), bytes4(keccak256("exec()"))));
        authed.exec();

        roles.setPublicAction(address(authed), bytes4(keccak256("exec()")), false);
        assertEq(roles.publicActions(address(authed), bytes4(keccak256("exec()"))), 0);
        assertTrue(!roles.canCall(address(this), address(authed), bytes4(keccak256("exec()"))));
        vm.expectRevert("AuthedMock/not-authorized");
        authed.exec();
    }
```

File: src/AllocatorRoles.sol [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+    constructor() {
+        wards[msg.sender] = 1;
+        emit Rely(msg.sender);
+    }
+
+    // --- getters ---
+
+    function hasUserRole(address who, uint8 role) external view returns (bool) {
+        return bytes32(0) != userRoles[who] & bytes32(uint256(uint256(2) ** uint256(role)));
```
Comment (telome) [inline src/AllocatorRoles.sol]: Is the middle cast necessary?

```suggestion
        return bytes32(0) != userRoles[who] & bytes32(uint256(2) ** uint256(role));
```

File: src/AllocatorRoles.sol [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+        emit Rely(usr);
+    }
+
+    function deny(address usr) external auth {
+        wards[usr] = 0;
+        emit Deny(usr);
+    }
+
+    function setUserRole(address who, uint8 role, bool enabled) public auth {
+        bytes32 mask = bytes32(uint256(uint256(2) ** uint256(role)));
```
Comment (telome) [inline src/AllocatorRoles.sol]: same here 
```suggestion
        bytes32 mask = bytes32(uint256(2) ** uint256(role));
```

File: src/AllocatorRoles.sol [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+        emit SetUserRole(who, role, enabled);
+    }
+
+    function setPublicAction(address target, bytes4 sig, bool enabled) external auth {
+        publicActions[target][sig] = enabled ? 1 : 0;
+        emit SetPublicAction(target, sig, enabled);
+    }
+
+    function setRoleAction(uint8 role, address target, bytes4 sig, bool enabled) external auth {
+        bytes32 mask = bytes32(uint256(uint256(2) ** uint256(role)));
```
Comment (telome) [inline src/AllocatorRoles.sol]:
```suggestion
        bytes32 mask = bytes32(uint256(2) ** uint256(role));
```

File: test/AllocatorRoles.t.sol (hunk 1 above)
Comment (sunbreak1211) [reply test/AllocatorRoles.t.sol]: Fixed!

File: src/AllocatorRoles.sol (hunk 2 above)
Comment (sunbreak1211) [reply src/AllocatorRoles.sol]: I actually simplified it a bit further.

File: src/AllocatorRoles.sol (hunk 3 above)
Comment (sunbreak1211) [reply src/AllocatorRoles.sol]: Same!

File: src/AllocatorRoles.sol (hunk 4 above)
Comment (sunbreak1211) [reply src/AllocatorRoles.sol]: Same!

**Review by telome — APPROVED** [review]
Looks good to merge to dev

**Pull request merged**